### PR TITLE
'Large Party Map' Performance - Part 1/x

### PR DESCRIPTION
**Pull request by 0xdevalias**
**Note:** this is only the first PR for _'Large Party Map' Performance_. I'm splitting them up to make code review easier, and to reduce the risk of deployments by working incrementally. I will link the others here when they are also ready.

----

**IMPORTANT:** This code should be cloned/checked by running before considering it ready to be merged. Subtle breakages may occur due to fixing re-renders (eg. where a feature was inadvertently dependent on the buggy re-render logic)

----

## Summary

- add firebase helper scripts
- add `useUserLastSeenLimit` (extracted from `useConnectPartyGoers`)
- add selectors
  - 'core' selectors for firestore: `firestoreDataSelector`, `firestoreOrderedSelector`, `firestoreStatusSelector`
  - `authSelector`, `profileSelector`, `currentVenueSelector`, `partygoersSelector`, `currentEventSelector`, `userPurchaseHistorySelector`, `shouldRetainAttendanceSelector`, `isCurrentVenueRequestedSelector`, `isCurrentEventRequestedSelector`, `isUserPurchaseHistoryRequestedSelector `
- add 'core' 'make selector' helpers for firestore
- refactor `Firestore` types to separately export `FirestoreStatus`, `FirestoreData`, `FirestoreOrdered` types + sort keys alphabetically + add `ValidFirestoreKeys` type
- add `useSparkleFirestoreConnect()` hook that uses `ValidFirestoreKeys` to constrain the allowed values for `storeAs`
- refactor `useConnectPartyGoers()` to also export selected `partygoers` data
- refactor a number of components to replace inline selectors with external 'static reference' selectors
  - including `VenuePage`, etc
- refactor a number of components/hooks/helpers/etc to use named exports + deprecate their `default` exports
  - note: refactoring all usages of the default exports was considered OOS/tech debt for future cleanup
- cleanup imports on legacy `PartyMap` component + related components (`RoomModal`, `Map`, `PartyTitle`, `RoomAttendance`, `RoomCard`, `RoomList`, `RoomModalOngoingEvent`, `ScheduleItem`)
  - exported all named components from `src/components/templates/PartyMap/components/index.ts`
- cleanup imports on `VenuePage`
- probably a bunch more things now... check the commits for a full list..

You can also look at the [commits](https://github.com/sparkletown/sparkle/pull/721/commits), which should be fairly descriptive/self contained.

### CODE REVIEW

**Review by thecadams — APPROVED** [review]
LGTM, good cleanup!

**Comment by 0xdevalias** [comment]
@thecadams I moved this back to draft because it seems to break the app. I believe the old logic was relying on some of the buggy re-rendering behaviour that should now be fixed. Which leaves `VenuePage` spinning on the `Loading...` component.

Once I figure out the fixes to make it work properly I will un-draft mode this PR and submit it for review again.

File: src/hooks/useSparkleFirestoreConnect.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * through a deep comparison using lodash's isEqual() function.
+ *
+ * @param config the config to be passed to useFirestoreConnect()
+ *
+ * @see SparkleRFQConfig
+ * @see ValidFirestoreKeys
+ * @see ReduxFirestoreQuerySetting
+ */
+export const useSparkleFirestoreConnect = (config: SparkleRFQConfig[]) =>
+  useFirestoreConnect(config);
```
Comment (0xdevalias) [inline src/hooks/useSparkleFirestoreConnect.ts:37]: Calling these out so people are aware of them, they should probably be used as our canonical go to replacement for `useFirestoreConnect()`

File: src/types/Firestore.ts [hunk 2]
```diff
@@ -1,88 +1,90 @@
...
+  data: FirestoreData;
+  ordered: FirestoreOrdered;
+  status: FirestoreStatus;
+}
+
+export interface FirestoreStatus {
+  requesting: Record<ValidFirestoreKeys, boolean>;
+  requested: Record<ValidFirestoreKeys, boolean>;
+  timestamps: Record<ValidFirestoreKeys, number>;
+}
```
Comment (0xdevalias) [inline src/types/Firestore.ts:43]: Note that these types are now automagically constructed for us, so long as we correctly add an entry to both `FirestoreData` and `FirestoreOrdered`.

File: src/types/Firestore.ts [hunk 3]
```diff
@@ -1,88 +1,90 @@
...
+
+// note: these entries should be sorted alphabetically
+export interface FirestoreData {
```
Comment (0xdevalias) [inline src/types/Firestore.ts:46]: Keeping these sorted alphabetically should make it easier when looking for keys

File: src/types/Firestore.ts [hunk 4]
```diff
@@ -1,88 +1,90 @@
...
+  userPurchaseHistory: Record<string, Purchase>;
+  userRoles: Record<string, Role>;
+  users: Record<string, User>;
+  venueChats: Record<string, RestrictedChatMessage> | null;
+  venueEvents: Record<string, VenueEvent>;
+  venues?: Record<string, AnyVenue>;
+}
+
+// note: these entries should be sorted alphabetically
+export interface FirestoreOrdered {
```
Comment (0xdevalias) [inline src/types/Firestore.ts:70]: Keeping these sorted alphabetically should make it easier when looking for keys

**Review by 0xdevalias — COMMENTED** [review]
Calling out some of the new additions I made + patterns they enable.

**Comment by 0xdevalias** [comment]
Ok, I believe this is fixed now. The commit that touched the most things, and that I haven't had the chance to go through and fully test all of the code paths for is https://github.com/sparkletown/sparkle/pull/721/commits/f7798176b0d77c28653247e15ed8466815cf2c10

Basically, in using `useConnectCurrentVenueNG()` to fix the some of the firestore connection, I had to go through and update any of the selectors that were expecting the data to be there from the old connect.

I think the first step I would take in testing this is to check that all of the updated components are used as children to a component that uses `useConnectCurrentVenueNG()` (currently that should only be `VenuePage`), or that those components that use `useConnectCurrentVenueNG()` have been rendered before them.

The next 'real world test' step to be certain would be to figure out all of the code paths where one of those components is rendered, and to access a page that will hit it in the web browser, and check that it all still works as expected.

I can look at doing this tomorrow if noone else is able to beforehand.

----

A future improvement to this pattern would be to use `createContext` and connect them that way + some nicer output if the context isn't available, but OOS'd it for the time being (at least in this PR)

File: src/pages/FriendShipPage/index.ts [hunk 5]
```diff
@@ -1 +1,6 @@
+export { FriendShipPage } from "./FriendShipPage";
```
Comment (denisdimitrov) [inline src/pages/FriendShipPage/index.ts:1]: Usually named exports are being used when you need to import multiple values from a single file. In this case we always have a single component. It's almost like you are destructuring an object with only 1 key. Just pointing out.

**Review by denisdimitrov — APPROVED** [review]
LGTM

File: src/pages/FriendShipPage/index.ts (hunk 5 above)
Comment (0xdevalias) [reply src/pages/FriendShipPage/index.ts:1]: _nods_ this was an explicit choice. I've read a few things over the years that suggest using default exports leads to more potential issues/etc. A few links from a quick skim of google:

- https://basarat.gitbook.io/typescript/main-1/defaultisbad
- https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad
- https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/

Obviously this speaks to a stylistic/etc choice, and we should look at the evidence for/against as a team (when we have time) and decide what we want our canonical choice here to be.

**Pull request merged**